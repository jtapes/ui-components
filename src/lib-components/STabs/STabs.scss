@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-tabs {
  &__header {
    position: relative;
    display: flex;
    border-bottom: 1px solid $color--border;

    &-item {
      position: relative;
      padding-bottom: $size--12;
      color: $color--font-secondary;
      transition: color ease $animate--time;
      cursor: pointer;

      &:not(:first-child) {
        margin-left: $size--80;
      }

      &:not(&.s-tabs__header-item--active):hover {
        color: $color--danger;
      }

      &--active {
        color: $color--primary;
        cursor: default;
        pointer-events: none;
      }
    }
  }

  &__content {
    display: flex;
  }

  &__slider {
    position: absolute;
    bottom: -($size--4);
    width: 0;
    height: $size--4;
    background-color: $color--danger;
    transition-property: left, width;
    transition-duration: $animate--time;
  }
}
