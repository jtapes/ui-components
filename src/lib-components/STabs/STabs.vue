<template>
  <div class="s-tabs">
    <div class="s-tabs__header">
      <div
        v-for="(tab, index) in tabs"
        :key="index"
        :ref="value === tab ? 'activeTab' : null"
        class="s-tabs__header-item bold h4"
        :class="{ 's-tabs__header-item--active': value === tab }"
        @click="handleValueChange(tab)"
      >
        {{ tab }}
      </div>
      <div v-if="tabs.length" ref="slider" class="s-tabs__slider" />
    </div>
    <div class="s-tabs__content">
      <template v-for="(tab, index) in tabs">
        <!-- @slot Слот для контента конкретного таба (tab-1, tab-2 и т.д.) -->
        <slot v-if="value === tab" :name="`tab-${index + 1}`" />
      </template>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model } from 'vue-property-decorator'

/** Компонент используется для отображения информации с возможностью переключать контентный блок */
@Component
export default class STabs extends Vue {
  /** Массив с названиями табов */
  @Prop({ type: Array, default: () => [] })
  tabs!: string[]

  /** Переменная, содержащая название активного таба */
  @Model('change')
  value!: string

  $refs!: {
    activeTab: HTMLDivElement[]
    slider: HTMLDivElement
  }

  handleValueChange(tab: string): void {
    if (this.value !== tab) {
      this.$emit('change', tab)
      this.$nextTick(() => {
        this.setSliderStyles()
      })
    }
  }

  setSliderStyles(): void {
    if (this.tabs.length && this.$refs.slider) {
      this.$refs.slider.style.width = `${this.$refs.activeTab[0].clientWidth}px`
      this.$refs.slider.style.left = `${this.$refs.activeTab[0].offsetLeft}px`
    }
  }

  mounted(): void {
    this.setSliderStyles()
  }
}
</script>
