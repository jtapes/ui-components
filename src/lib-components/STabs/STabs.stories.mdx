import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Preview,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import STabs from '@/lib-components/STabs/STabs.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Tabs"
  component={STabs}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    tabs: { control: 'array', defaultValue: ['Пункт 1', 'Пункт 2', 'Пункт 3'] },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=3019%3A1571',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Tabs

<Description of={STabs} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { STabs },
  template:
    `<s-tabs v-model="activeTab" :tabs="tabs">
         <template #tab-1>
             <span class="mt-40">111</span>
         </template>
         <template #tab-2>
             <span>222</span>
         </template>
         <template #tab-3>
             <span>333</span>
         </template>
     </s-tabs>`,
  data: () => ({
    activeTab: 'Пункт 1',
  })
})

<Preview withToolbar withSource="none">
  <Story name="STabs">
    {Template.bind({})}
  </Story>
</Preview>

## Пример
<Source
  language="html"
  code={`
<s-tabs :tabs="tabs">
  <template #tab-1>
    <span>123</span>
  </template>
  <template #tab-3>
    <span class="mt-40">321</span>
  </template>
</s-tabs>
  `}
/>

<div>Слоты именуются, как: tab-n, где "n" — это номер таба (начинается с 1).</div>

<ArgsTable of={STabs} />
