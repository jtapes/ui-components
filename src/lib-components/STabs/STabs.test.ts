import { mount } from '@vue/test-utils'
import STabs from '@/lib-components/STabs/STabs.vue'

describe('STabs', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount({
      template: `
        <s-tabs v-model="activeTab" :tabs="tabs">
          <template #tab-1>
            <span class="mt-40">111</span>
          </template>
          <template #tab-2>
              <span>222</span>
          </template>
          <template #tab-3>
              <span>333</span>
          </template>
        </s-tabs>`,
      components: { STabs },
      data: () => ({
        activeTab: 'Пункт 1',
        tabs: ['Пункт 1', 'Пункт 2', 'Пункт 3'],
      }),
    })
    expect(wrapper.element).toMatchSnapshot()
  })
  test('Is emitted by $emit', async () => {
    const wrapper = mount(STabs, {
      propsData: {
        value: 'Пункт 1',
        tabs: ['Пункт 1', 'Пункт 2', 'Пункт 3'],
      },
    })

    const nextTab = wrapper.findAll('.s-tabs__header-item').wrappers[1]
    await nextTab.trigger('click')

    expect(wrapper.emitted('change')).toBeTruthy()
  })
  test('Test with empty props', () => {
    const wrapper = mount(STabs)

    expect(wrapper.vm.$props.tabs.length).toBe(0)
  })
})
