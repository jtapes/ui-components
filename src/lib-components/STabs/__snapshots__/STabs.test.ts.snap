// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`STabs Is Vue snapshot 1`] = `
<div
  class="s-tabs"
>
  <div
    class="s-tabs__header"
  >
    <div
      class="s-tabs__header-item bold h4 s-tabs__header-item--active"
    >
      
      Пункт 1
    
    </div>
    <div
      class="s-tabs__header-item bold h4"
    >
      
      Пункт 2
    
    </div>
    <div
      class="s-tabs__header-item bold h4"
    >
      
      Пункт 3
    
    </div>
     
    <div
      class="s-tabs__slider"
      style="width: 0px; left: 0px;"
    />
  </div>
   
  <div
    class="s-tabs__content"
  >
    <span
      class="mt-40"
    >
      111
    </span>
    <!---->
    <!---->
  </div>
</div>
`;
